test/DaedalusCompiler.Tests/AssemblyBuilderTest.cs

```diff
@@ -0,0 +1,181 @@
+﻿using System.Collections.Generic;
+using Antlr4.Runtime;
+using Antlr4.Runtime.Tree;
+using DaedalusCompiler.Compilation;
+using DaedalusCompiler.Dat;
+using Xunit;
+
+namespace DaedalusCompiler.Tests
+{
+    public class AssemblyBuilderTest
+    {
+        private static readonly Dictionary<string, DatSymbolType> StringToTypeMap =
+            new Dictionary<string, DatSymbolType>
+            {
+                {"void", DatSymbolType.Void},
+                {"float", DatSymbolType.Float},
+                {"int", DatSymbolType.Int},
+                {"string", DatSymbolType.String},
+                {"class", DatSymbolType.Class},
+                {"func", DatSymbolType.Func},
+                {"prototype", DatSymbolType.Prototype},
+                {"instance", DatSymbolType.Instance},
+            };
+
+        private DatSymbol Var(string variable)
+        {
+            string[] typeAndName = variable.Split(' ');
+            return SymbolBuilder.BuildVariable(typeAndName[1], StringToTypeMap[typeAndName[0]]);
+        }
+
+        private DatSymbol Func(string funcname)
+        {
+            return SymbolBuilder.BuildFunc(funcname, DatSymbolType.Func);
+        }
+
+        private List<AssemblyElement> ParseExpressions(string declarations, string expressions)
+        {
+            string data = $@"
+                {declarations}
+                func void testFunc()
+                {{
+                    {expressions}
+                }};
+            ";
+            AntlrInputStream inputStream = new AntlrInputStream(data);
+            DaedalusLexer lexer = new DaedalusLexer(inputStream);
+            CommonTokenStream commonTokenStream = new CommonTokenStream(lexer);
+            DaedalusParser parser = new DaedalusParser(commonTokenStream);
+
+
+            AssemblyBuilder assemblyBuilder = new AssemblyBuilder();
+            ParseTreeWalker.Default.Walk(new DaedalusParserListener(assemblyBuilder, 0), parser.daedalusFile());
+            return assemblyBuilder.functions.Find(func => func.symbol.Name == "testFunc").body;
+        }
+
+        private void CompareInstructionLists(
+            List<AssemblyElement> instructions,
+            List<AssemblyElement> expectedInstructions)
+        {
+            for (var index = 0; index < expectedInstructions.Count; index++)
+            {
+                var instruction = instructions[index];
+                var expectedInstruction = expectedInstructions[index];
+                CompareInstructions(instruction, expectedInstruction);
+            }
+        }
+
+        private void CompareInstructions(AssemblyElement instruction, AssemblyElement expectedInstruction)
+        {
+            Assert.Equal(instruction.GetType(), expectedInstruction.GetType());
+            switch (instruction)
+            {
+                case PushInt _:
+                {
+                    Assert.Equal(
```

Comment:
Are you sure that casting for `instruction` is needed ? You can use something like `case PushInt pushIntElem` and casting is not needed, casting for expectedInstruction is ok,
the same for bellow cases